src/legacy/ui/public/state_management/state_hashing/hash_unhash_url.ts

```diff
@@ -0,0 +1,157 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import rison, { RisonObject } from 'rison-node';
+import { stringify as stringifyQueryString } from 'querystring';
+import encodeUriQuery from 'encode-uri-query';
+import { format as formatUrl, parse as parseUrl } from 'url';
+import { HashedItemStoreSingleton } from '../state_storage';
+import { createStateHash, isStateHash } from './state_hash';
+
+export type IQuery = Record<string, any>;
+
+interface QueryMapperOptions {
+  hashableParams: string[];
+}
+export type QueryReplacerOptions = QueryMapperOptions;
+
+export const unhashQuery = createQueryMapper(stateHashToRisonState);
+export const hashQuery = createQueryMapper(risonStateToStateHash);
+
+export const unhashUrl = createQueryReplacer(unhashQuery);
+export const hashUrl = createQueryReplacer(hashQuery);
+
+// naive hack, but this allows to decouple these utils from AppState, GlobalState for now
+// when removing AppState, GlobalState and migrating to IState containers,
+// need to make sure that apps explicitly passing this whitelist to hash
+const __HACK_HARDCODED_LEGACY_HASHABLE_PARAMS = ['_g', '_a'];
```

Comment:
This is probably the most controversial thing and I am really not sure about.
This allows to decouple this utils from State objects and I hope it will go away when we start wiring it up to state containers.

Reply:
I'm noticing we also have a fallback `_s` in State, which appears to never be used. (I think)

Let me think on this a bit more and maybe we can discuss further, but my initial reaction is that this is probably fine -- the applications themselves will need to provide these params in the new platform, and therefore they will need to be configurable in the utilities we provide anyway. So taking a step like this would only be a temporary measure until we have the apps using the new utilities, at which point we could remove the hardcoded params and use whatever is passed to the util.

Reply:
Let's discuss it further, but that is exactly how I was looking at it so far :+1

Reply:
I added '_s' as it was used in state.js unit tests.

Reply:
Yeah now this makes sense based on our discussion around how these keys will be configured in the new syncing utility.  👍